### PR TITLE
Action Service & Factory

### PR DESCRIPTION
**Pull request by Minh-Than**
The sole reason for this to exist is simply because I wanted to learn writing them since I'm really not a fan of seeing voodoo magic in the lines of code and not understanding absolute zilch. This took about 4? 6? days of constant mental breakdowns due to lack of knowledge and shallower learning capability than a pool of fishes that eat your extraneous feet bits to promote healthy skin. 

Another reason I definitely did not think of before is that this reduces the amount of `addActionListeners()` that would have to be shuffled around, will help a lot when the major UI change arrives because dealing with a lot less `addActionListeners()` will be less painful than dealing with a hundred of them.

Anyway, I wanted to have an action "service" which its main purpose is to register actions, specifically the codes that would run when something triggers the action, similar to how `ButtonService` uses `registerButton()` or `registerLabel` to link `JButton` and `JLabel` respectively to link them to actions. Initially. This service contains `registeredActions` which initially will be registered with the existing Action classes that don't really fit into this service/couldn't make me be bothered enough to look at ('FoldAction', `DuplicateFoldedModelAction`, ...) in it. It's then all appended with the rest of actions and their codes during the start up. The default Action class to use is `LambdaAction`, which simply takes an `ActionType` and a `Runnable` to accept a function. Some actions have similar steps with different variables so some template Action classes are made for those as well.

An example would be: 
`actionService.registerAction(new LamdbaAction(canvasModel, mainCreasePatternWorker, Action.flushAction, () -> flushToilet()));`

Some improvements (-ish) were made after the fact to prevent the need to type in existing models and services (`CanvasModel`, `AngleSystemModel` and potentially more) when registering the actions, namely creating an Action factory, which injects those existing models into the methods which then returns the aforementioned template classes.

An example would be: 
`actionService.registerAction(actionFactory.setMouseMode(Action.[...], () -> MouseMode.[...];`

The above are loosely paraphrased by words from trox because what did you expect from me at this point. Did I learn anything from this? I mean I know something, not close to being as much as I'd hope to learn from the breakdowns I mentioned. I'll have to write more of this, bonus points of it's also more useful.

### CODE REVIEW

**Comment by Minh-Than** [comment]
And reviewers please take your time to check everything if you can. There's so many things that could've gone wrong and I maybe didn't even realize and this is the biggest PR I don't think I would make.

File: oriedita/src/main/java/oriedita/editor/service/impl/ActionServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        extractActionInstances(actions);
+    }
+
+    @Override
+    public synchronized void registerAction(OrieditaAction orieditaAction) { registeredActions.add(orieditaAction); }
+
+    @Override
+    public List<OrieditaAction> getAllRegisteredActions() { return registeredActions; }
+
+    private void extractActionInstances(Instance<OrieditaAction> actions){ actions.stream().forEach(registeredActions::add); }
```
Comment (undertrox) [inline oriedita/src/main/java/oriedita/editor/service/impl/ActionServiceImpl.java:31]: this::registerAction would be better here. Even though they currently do the same thing, if the implementation of registerAction ever changes, you shouldnt need to update this method too

Comment (Minh-Than) [reply oriedita/src/main/java/oriedita/editor/service/impl/ActionServiceImpl.java]: Done, hopefully correctly

**Comment by Minh-Than** [comment]
One thing I should mention is that I don't really mind if this PR is rejected. Main goal of this PR is to have something I can remind myself if I need to check how to write service/factory.

And it's kind of preventing some other nifty thing I want to add in since I'm not wild with having to deal with conflicts.

But for now I will mark this as draft.

**Pull request merged**